reference-implementation/to-upstream-wpts/writable-streams/aborting.js

```diff
@@ -402,56 +405,59 @@ promise_test(() => {
     writer.close();
     const abortPromise = writer.abort();
     return flushAsyncEvents().then(() => {
-      assert_false(abort_called, 'abort should not be called while close is pending');
+      assert_array_equals(ws.events, ['close'], 'abort should not be called while close is pending');
       resolveClose();
-      return abortPromise.then(() => assert_false(abort_called, 'abort should not be called after close completes'));
+      return abortPromise.then(() => {
+        assert_array_equals(ws.events, ['close'], 'abort should not be called while close is pending');
+      });
     });
   });
 }, 'underlying abort() should not be called if underlying close() has started');
 
 promise_test(t => {
   let resolveWrite;
-  let abort_called = false;
-  const ws = new WritableStream({
+  const ws = recordingWritableStream({
     write() {
       return new Promise(resolve => {
         resolveWrite = resolve;
       });
-    },
-    abort() {
-      abort_called = true;
     }
   });
 
   const writer = ws.getWriter();
   return writer.ready.then(() => {
     writer.write('a');
     const closePromise = writer.close();
-    const abortPromise = writer.abort();
+    const abortPromise = writer.abort('b');
+
     return flushAsyncEvents().then(() => {
-      assert_false(abort_called, 'abort should not be called while write is pending');
+      assert_array_equals(ws.events, ['write', 'a'], 'abort should not be called while write is pending');
       resolveWrite();
       return abortPromise.then(() => {
-        assert_true(abort_called, 'abort should be called after write completes');
+        assert_array_equals(ws.events, ['write', 'a', 'abort', 'b'], 'abort should be called after write completes');
         return promise_rejects(t, new TypeError(), closePromise, 'promise returned by close() should be rejected');
       });
     });
   });
-}, 'underlying abort() should be called while closing if underlying close() has not started yet');
+}, 'an abort() that happens during a write() should trigger the underlying abort() even with a close() queued');
 
 promise_test(() => {
   const ws = new WritableStream();
   const writer = ws.getWriter();
   return writer.ready.then(() => {
     const closePromise = writer.close();
     const abortPromise = writer.abort();
-    let closeResolved = false;
-    Promise.all([
-      closePromise.then(() => { closeResolved = true; }),
-      abortPromise.then(() => { assert_true(closeResolved, 'close() promise should resolve before abort() promise'); })
-    ]);
+    const ordering = [];
+
+    return Promise.all([
+      closePromise.then(() => { ordering.push('close'); }),
+      abortPromise.then(() => { ordering.push('abort'); })
+    ])
+    .then(() => {
+      assert_array_equals(ordering, ['abort', 'close']);
+    });
   });
-}, 'writer close() promise should resolve before abort() promise');
+}, 'writer abort() promise should fulfill before close() promise (since it bails early)');
```

Comment:
I'm not keen on this behaviour. Usually we can rely on abort() to tell us whether the stream was cleaned up or not, but in this case we can't.